sw/dnn/layernorm/src/layernorm.h

```diff
@@ -246,26 +250,493 @@ static inline void layernorm_fp32_opt(float *input, float *output,
+                    "vfcpkb.b.s %[mean_reg], %[var_v2_0], %[var_v2_0] \n"
+                    "vfcpkc.b.s %[mean_reg], %[var_v2_0], %[var_v2_0] \n"
+                    "vfcpkd.b.s %[mean_reg], %[var_v2_0], %[var_v2_0] \n"
+                    : [ mean_v2_0 ] "+f"(mean_v2[0].f64),
+                      [ mean_v2_1 ] "+f"(mean_v2[1].f64),
+                      [ mean_v2_2 ] "+f"(mean_v2[2].f64),
+                      [ mean_v2_3 ] "+f"(mean_v2[3].f64),
+                      [ mean_v4_0 ] "+f"(mean_v4[0].f64),
+                      [ mean_v4_1 ] "+f"(mean_v4[1].f64),
+                      [ mean_v4_2 ] "+f"(mean_v4[2].f64),
+                      [ mean_v4_3 ] "+f"(mean_v4[3].f64),
+                      [ var_v8_0 ] "+f"(var_v8[0].f64),
+                      [ var_v8_1 ] "+f"(var_v8[1].f64),
+                      [ var_v8_2 ] "+f"(var_v8[2].f64),
+                      [ var_v8_3 ] "+f"(var_v8[3].f64),
+                      [ mean_reg ] "+f"(mean_reg.f64),
+                      [ pow_v8_0 ] "+f"(pow_v8[0].f64),
+                      [ pow_v8_1 ] "+f"(pow_v8[1].f64),
+                      [ pow_v8_2 ] "+f"(pow_v8[2].f64),
+                      [ pow_v8_3 ] "+f"(pow_v8[3].f64),
+                      [ var_v2_0 ] "+f"(var_v2[0].f64),
+                      [ var_v2_1 ] "+f"(var_v2[1].f64),
+                      [ var_v2_2 ] "+f"(var_v2[2].f64),
+                      [ var_v2_3 ] "+f"(var_v2[3].f64)
+                    : [ zero ] "f"(0.0f), [ n_frep ] "r"(n_frep - 1),
+                      [ embeddings ] "f"((float)embeddings),
+                      [ eps ] "f"((float)eps), [ one ] "f"(1.0f)
+                    : "ft0", "ft1", "ft2");
+
+                snrt_fpu_fence();
+
+                snrt_ssr_read(SNRT_SSR_DM2, SNRT_SSR_2D,
+                              &core_otile[b * batch_offset + s * stride]);
+
+                asm volatile(
+                    "frep.o  %[n_frep], 4, 0, 0 \n"
+                    "vfmul.b ft1, ft2, %[mean_reg] \n"
+                    "vfmul.b ft1, ft2, %[mean_reg] \n"
+                    "vfmul.b ft1, ft2, %[mean_reg] \n"
+                    "vfmul.b ft1, ft2, %[mean_reg] \n"
+                    : [ mean_reg ] "+f"(mean_reg.f64)
+                    : [ n_frep ] "r"(n_frep - 1)
+                    : "ft0", "ft1", "ft2");
+                snrt_ssr_disable();
+            }
+        }
+    }
+}
 // Tiles the seq_len axis (assumes seq_len is an integer multiple of n_tiles)
 // Distributes tiles to clusters (assumes n_tiles is an integer multiple of
 // the number of clusters)
 static inline void layernorm_layer(layernorm_layer_t l) {
+    uint32_t data_type_size;
+    if (l.dtype == FP32) {
+        data_type_size = sizeof(float);
+    } else if (l.dtype == FP16) {
+        data_type_size = sizeof(__fp16);
+    } else if (l.dtype == FP8) {
+        data_type_size = sizeof(char);
+    } else {
+        printf("Unsupported data type\n");
+    }
```

Comment:
The value of the enumeration type is actually already the `data_type_size`, so you don't need this here. See https://github.com/pulp-platform/snitch_cluster/blob/04f80a17905d06774d378ad9c932968d5c790090/sw/dnn/src/dnn.h#L13